buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy

```diff
@@ -2063,8 +2066,13 @@ class BeamModulePlugin implements Plugin<Project> {
               args '-c', ". ${project.ext.envdir}/bin/activate && cd ${copiedPyRoot} && scripts/run_tox.sh $tox_env $distTarBall '$posargs'"
             }
           }
-          inputs.files project.pythonSdkDeps
-          outputs.files project.fileTree(dir: "${pythonRootDir}/target/.tox/${tox_env}/log/")
+          outputs.cacheIf { true }
+          inputs.files(project.pythonSdkDeps)
+              .withPropertyName('pythonSdkDeps')
+              .withPathSensitivity(PathSensitivity.RELATIVE)
+
+          outputs.dir("${project.buildDir}/srcs/sdks/python/")
```

Comment:
Switching to `$project.buildDir` is definitely a major improvement. Is there a reason that tox outputs are dumped alongside the source? Is that something we can easily control and put them in, say, `build/reports/tox` or some such?

Reply:
I don't have an alternative solution, but it seems that marking everything under `sdks/python` as output will include the output from all `tox_envs`.

Reply:
Just to clarify the full path, it is marking everything under `build/src/sdks/python` as output here. Am I wrong that `${project.buildDir}` is the `build/` directory relative to the `build.gradle` file of the module? Certainly everything under there is fine to be an output. But keeping outputs from different tasks in totally different places is nice. Just a nice to have. So `build/src/sdks/python` is the copy/pasted source while `build/reports/tox/` would be the output from tox. But TBH I am just dropping in with generic opinions. My very quick read of the build logic and comments was that `build/src/sdks/python` existed primarily so that tox (and/or other tools) would not dump files all over the original source tree. So if that is its purpose it makes sense to let the tools put their stuff in their and call it good.

Reply:
You're right, according to the docs:
projectDir | File | The directory containing the build script.
-- | -- | --
buildDir | File | projectDir/build

https://docs.gradle.org/current/userguide/writing_build_scripts.html
(aside: copy-pasting a table just generates the markdown? what)

So forget my comment above. However, by my understanding the new definition of `outputs` is still going to be empty after the task is run, but the screenshot above shows it has a copy of the repo (`pythonSdkDeps`?). What's causing this?

Reply:
I'm not sure I understand.

I think by this line `outputs` should contain a dir-type output for `build/srcs/sdks/python`:

```python
          outputs.dir("${project.buildDir}/srcs/sdks/python/")
```

The screenshot shows the state of `build/`, which isn't limited to the things in `outputs`. At least part of it is explained by:

```python
      def copiedSrcRoot = "${project.buildDir}/srcs"
      ...
     project.copy { from project.pythonSdkDeps; into copiedSrcRoot }
```

If I were debugging this, I would start by making tasks have more disjoint directories named roughly after the task itself so I was very certain that each directory had sole ownership and I could easily map the directory back to the task that created it.